### PR TITLE
RUBY-2119 Do not lose type information when writing documents with automatic encryption

### PR DESCRIPTION
**Pull request by egiurleo**
**The Problem:** During automatic encryption, commands are sent to mongocryptd to have encryption placeholders added to them. The response from mongocryptd is deserialized without the `mode: :bson` flag, causing command fields that were previously BSON types (e.g. Int64s) to be deserialized into Ruby types (e.g. Ruby integers). This means that the driver can lose type information on non-encrypted fields when writing them to the database using automatic encryption.

**The Fix:** Allow commands to pass an option called `deserialize_as_bson` to the execution code and down to the protocol layer, which will then deserialize responses from the server using the `mode: :bson` option. This isn't included on all commands because it would create a significant change in driver behavior, so it is only used on the command that is sent to mongocryptd.

### CODE REVIEW

File: lib/mongo/protocol/serializers.rb [hunk 1]
```diff
@@ -265,7 +265,7 @@ def self.serialize(buffer, value, max_bson_size = nil, validating_keys = BSON::C
           #
           # @since 2.5.0
           def self.deserialize(buffer)
-            BSON::Document.from_bson(buffer)
+            BSON::Document.from_bson(buffer, mode: :bson)
```
Comment (egiurleo) [inline lib/mongo/protocol/serializers.rb:268]: The problem was that we were sending commands to mongocryptd, which adds encryption placeholders, and then we were deserializing the responses without `mode: :bson`, so any int64s were coming back as regular integers.

File: lib/mongo/database.rb [hunk 2]
```diff
@@ -162,9 +162,17 @@ def collections
...
+    # @option opts :execution_options [ Hash ] Options to pass to the code that
+    #   executes this command. This is an internal option and is subject to
+    #   change.
+    #   - :deserialize_as_bson [ Boolean ] Whether to deserialize the response
+    #     to this command using BSON types intead of native Ruby types wherever
+    #     possible.
     #
     # @return [ Mongo::Operation::Result ] The result of the command execution.
     def command(operation, opts = {})
+      execution_opts = opts.delete(:execution_options) || {}
```
Comment (p-mongo) [inline lib/mongo/database.rb:174]: `opts` need to be duplicated before being mutated.

File: spec/integration/crud_spec.rb [hunk 3]
```diff
@@ -7,6 +7,83 @@
...
   end
 
+  describe 'insert' do
+    context 'inserting a BSON::Int64 or BSON::Int32' do
+      before do
+        collection.insert_one(int64: BSON::Int64.new(42))
+      end
+
+      it 'is stored as the correct type' do
+        result = collection.find(int64: { '$type' => 18 }).first
```
Comment (p-mongo) [inline spec/integration/crud_spec.rb:17]: I believe `'$type' => 'long'` would work here and be more clear.

Comment (p-mongo) [reply spec/integration/crud_spec.rb:17]: Per the docs at https://docs.mongodb.com/manual/reference/operator/query/type/ and https://docs.mongodb.com/manual/reference/operator/query/type/#document-type-available-types.

File: lib/mongo/database.rb [hunk 4]
```diff
@@ -171,18 +171,19 @@ def collections
     #
     # @return [ Mongo::Operation::Result ] The result of the command execution.
     def command(operation, opts = {})
-      execution_opts = opts.delete(:execution_options) || {}
+      opts_copy = opts.dup
```
Comment (p-mongo) [inline lib/mongo/database.rb:174]: This can simply be `opts = opts.dup` since generally the original argument won't be used past this point.

**Pull request merged**